File: solutions/0860. Lemonade Change/0860.py
# Time complexity: O(⁡n)
# Space complexity: O(1)
class Solution:
  def lemonadeChange(self, bills: List[int]) -> bool:
    count5 = 0
    count10 = 0

    for bill in bills:
      if bill == 5:
        count5 += 1
      elif bill == 10:
        count5 -= 1
        count10 += 1
      else:
        if count10 > 0:
          count10 -= 1
          count5 -= 1
        else:
          count5 -= 3
      if count5 < 0:
        return False

    return True

File: solutions/0860. Lemonade Change/0860.rs
// Time complexity: O(⁡n)
// Space complexity: O(1)
impl Solution {
    pub fn lemonade_change(bills: Vec<i32>) -> bool {
        let mut count5 = 0;
        let mut count10 = 0;

        for &bill in &bills {
            match bill {
                5 => count5 += 1,
                10 => {
                    if count5 > 0 {
                        count5 -= 1;
                        count10 += 1;
                    } else {
                        return false;
                    }
                }
                20 => {
                    if count10 > 0 && count5 > 0 {
                        count10 -= 1;
                        count5 -= 1;
                    } else if count5 >= 3 {
                        count5 -= 3;
                    } else {
                        return false;
                    }
                }
                _ => unreachable!(),
            }
        }

        true
    }
}
